cou/steps/analyze.py

```diff
@@ -0,0 +1,171 @@
+
+        return Analysis(apps=apps)
+
+    @classmethod
+    async def _populate(cls) -> set[Application]:
+        """Generate the applications model."""
+        juju_status = await async_get_full_juju_status()
+        model_name = juju_status.model.name
+        apps = {
+            Application(
+                name=app,
+                status=app_status,
+                config=await async_get_application_config(app),
+                model_name=model_name,
+            )
+            for app, app_status in juju_status.applications.items()
+        }
+        return apps
+
+    def __str__(self) -> str:
+        """Dump as string."""
+        return "\n".join([str(app) for app in self.apps])
+
+
+@dataclass
+class Application:
+    """Representation of an application in the deployment."""
+
+    # pylint: disable=too-many-instance-attributes
+
+    name: str
+    status: ApplicationStatus
+    config: dict
+    model_name: str
+    charm: str = ""
+    charm_origin: str = ""
+    os_origin: str = ""
+    channel: str = ""
+
+    # e.g: {"keystone/0": {'os_version': 'victoria', 'workload_version': '2:18.1'}}
+    units: defaultdict[str, dict] = field(default_factory=lambda: defaultdict(dict))
+
+    def __post_init__(self) -> None:
+        """Initialize the Application dataclass."""
+        self.charm = extract_charm_name_from_url(self.status.charm)
+        self.channel = self.status.charm_channel
+        self.charm_origin = self.status.charm.split(":")[0]
+        self.os_origin = self._get_os_origin()
+        for unit in self.status.units.keys():
+            workload_version = self._get_charm_workload_version(unit)
+            self.units[unit]["workload_version"] = workload_version
+            os_version = self._get_current_os_version(workload_version)
+            self.units[unit]["os_version"] = os_version
+
+    def __hash__(self) -> int:
+        """Hash magic method for Application."""
+        return hash(f"{self.name}{self.charm}")
+
+    def __eq__(self, other: Any) -> Any:
+        """Equal magic method for Application."""
+        return other.name == self.name and other.charm == self.charm
+
+    def __str__(self) -> str:
+        """Dump as string."""
+        summary = {
+            self.name: {
+                "model_name": self.model_name,
+                "charm": self.charm,
+                "charm_origin": self.charm_origin,
+                "os_origin": self.os_origin,
+                "channel": self.channel,
+                "units": {
+                    unit: {
+                        "workload_version": details.get("workload_version", ""),
+                        "os_version": details.get("os_version", ""),
+                    }
+                    for unit, details in self.units.items()
+                },
+            }
+        }
+        yaml = YAML()
+        with StringIO() as stream:
+            yaml.dump(summary, stream)
+            return stream.getvalue()
```

Comment:
Why we are not using yaml package instead of ruamel yaml? With yaml we can do `yaml.dump(summary)`

Reply:
hmm... I had suggested the switch because it looked like pyyaml was no longer being maintained, while ruamel.yaml is actively so. Now that I'm looking at this again I'm no longer sure. And ruamel's use of sourceforge is a turnoff.
There are lots of "pyyaml vs ruamel" or "switch to ruamel" github issues at various projects, but the discussions seem to come from a time when pyyaml was less active than today. And of course if all we need yaml for is dumping to string either is fine.

Reply:
To me ruamel.yaml less intuitive to use, but it's support newer yaml format. The PyYaml is easier to use and it's used often in our projects, but you are right it not well maintained. So both have some pros and cons, but I never check it properly. Your argument are correct, the PyYaml is older and not well maintained + Juju is using [gopkg.in/yaml.v2](https://pkg.go.dev/gopkg.in/yaml.v2?utm_source=godoc), which supports YAML 1.2.
Suggestion: Maybe we can create simple function to avoid using those "ugly " lines :D.
 
BTW: I should put "not-blocking" before me comment here.